File: server/lib/services/cleaning_service.dart
import 'package:minister_shared/models/transaction.dart';
import '../category_rules.dart';
import '../store/json_store.dart';

const _dropFields = [
  'object',
  'account',
  'livemode',
  'updated',
  'transaction_refresh',
  'transacted_at',
];

String categorize(
    String description, String transactionId, Map<String, String> overrides) {
  final override = overrides[transactionId];
  if (override != null) return override;

  for (final entry in categoryRules) {
    if (entry.key.hasMatch(description)) return entry.value;
  }
  return 'Uncategorized';
}

CleanTransaction cleanTransaction(
    StoredTransaction tx, Map<String, String> overrides) {
  final cleaned = <String, dynamic>{};

  for (final entry in tx.data.entries) {
    if (!_dropFields.contains(entry.key)) {
      cleaned[entry.key] = entry.value;
    }
  }

  cleaned['category'] = categorize(tx.description, tx.id, overrides);

  final transactedAt = tx.data['transacted_at'];
  if (transactedAt is int) {
    final d = DateTime.fromMillisecondsSinceEpoch(transactedAt * 1000, isUtc: true);
    cleaned['date'] = d.toIso8601String().substring(0, 10);
    cleaned['year'] = d.year;
    cleaned['month'] = d.month;
  }

  final st = tx.data['status_transitions'];
  if (st is Map) {
    final normalized = <String, dynamic>{};
    for (final entry in st.entries) {
      if (entry.value is int) {
        final d = DateTime.fromMillisecondsSinceEpoch(
            (entry.value as int) * 1000,
            isUtc: true);
        normalized[entry.key as String] = d.toIso8601String().substring(0, 10);
      } else {
        normalized[entry.key as String] = entry.value;
      }
    }
    cleaned['status_transitions'] = normalized;
  }

  return CleanTransaction(cleaned);
}

List<CleanTransaction> cleanAllTransactions() {
  final transactions = loadTransactions();
  final overrides = loadOverrides();

  final cleaned = transactions
      .where((tx) => tx.amount < 0)
      .map((tx) => cleanTransaction(tx, overrides))
      .toList();

  saveCleanTransactions(cleaned);
  return cleaned;
}

File: server/src/services/cleaningService.ts
import type { CategoryRule, CleanTransaction, StoredTransaction } from '../types.js';
import {
  loadCategoryRules,
  loadCleanTransactions,
  loadDefaultCategoryRules,
  loadOverrides,
  loadPinnedTransactions,
  loadTransactions,
  saveCleanTransactions,
} from '../store/jsonStore.js';

const DROP_FIELDS = new Set([
  'object',
  'account',
  'livemode',
  'updated',
  'transaction_refresh',
  'transacted_at',
]);

function buildRules(): Array<{ regexp: RegExp; category: string }> {
  const userRules = loadCategoryRules();
  const defaultRules = loadDefaultCategoryRules();
  // User rules first (higher precedence)
  return [...userRules, ...defaultRules].map((r: CategoryRule) => ({
    regexp: new RegExp(r.pattern, r.caseSensitive ? '' : 'i'),
    category: r.category,
  }));
}

export function categorize(
  description: string,
  transactionId: string,
  overrides: Record<string, string>,
): string {
  const override = overrides[transactionId];
  if (override) return override;

  const rules = buildRules();
  for (const rule of rules) {
    if (rule.regexp.test(description)) return rule.category;
  }
  return 'Uncategorized';
}

export function cleanTransaction(
  tx: StoredTransaction,
  overrides: Record<string, string>,
): CleanTransaction {
  const cleaned: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(tx)) {
    if (!DROP_FIELDS.has(key)) {
      cleaned[key] = value;
    }
  }

  // Normalize description
  if (typeof cleaned['description'] === 'string') {
    cleaned['description'] = cleaned['description'].replace(/\s+/g, ' ').trim();
  }

  const desc = (cleaned['description'] as string | undefined) ?? '';
  cleaned['category'] = categorize(desc, tx.id, overrides);

  const transactedAt = tx['transacted_at'];
  if (typeof transactedAt === 'number') {
    const d = new Date(transactedAt * 1000);
    cleaned['date'] = d.toISOString().slice(0, 10);
    cleaned['year'] = d.getUTCFullYear();
    cleaned['month'] = d.getUTCMonth() + 1;
  }

  const st = tx['status_transitions'];
  if (st && typeof st === 'object' && !Array.isArray(st)) {
    const normalized: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(st as Record<string, unknown>)) {
      if (typeof value === 'number') {
        normalized[key] = new Date(value * 1000).toISOString().slice(0, 10);
      } else {
        normalized[key] = value;
      }
    }
    cleaned['status_transitions'] = normalized;
  }

  return cleaned as CleanTransaction;
}

export function cleanAllTransactions(): CleanTransaction[] {
  const transactions = loadTransactions();
  const overrides = loadOverrides();
  const pinned = loadPinnedTransactions();

  // Load existing clean transactions to preserve categories for pinned ones
  const existingClean = new Map<string, CleanTransaction>();
  if (pinned.size > 0) {
    for (const tx of loadCleanTransactions()) {
      existingClean.set(tx.id, tx);
    }
  }

  const cleaned = transactions.map((tx) => {
    const result = cleanTransaction(tx, overrides);
    // If pinned, restore its existing category
    if (pinned.has(tx.id) && existingClean.has(tx.id)) {
      result.category = existingClean.get(tx.id)!.category;
    }
    return result;
  });

  saveCleanTransactions(cleaned);
  return cleaned;
}
